File: src/main/java/main/block/power/PowerSystem.java
package main.block.power;

import java.util.HashMap;
import java.util.Map;

import javax.annotation.Nullable;

import it.unimi.dsi.fastutil.longs.Long2ObjectMap;
import it.unimi.dsi.fastutil.longs.Long2ObjectOpenHashMap;
import main.SkylaExtension;
import net.minestom.server.instance.Instance;
import net.minestom.server.utils.BlockPosition;
import net.minestom.server.utils.chunk.ChunkUtils;

// Defines the behavior of the custom power system for one instance
public class PowerSystem {
	
	private static final Map<Instance, PowerSystem> powerSystems = new HashMap<>();
	
	private final Long2ObjectMap<PowerBlock[]> powerGrid = new Long2ObjectOpenHashMap<>();
	
	final Instance instance;
	
	public static PowerSystem of(Instance instance) {
		if (powerSystems.containsKey(instance))
			return powerSystems.get(instance);
		
		PowerSystem powerSystem = new PowerSystem(instance);
		
		powerSystems.put(instance, powerSystem);
		
		return powerSystem;
	}
	
	private PowerSystem(Instance instance) {
		this.instance = instance;
	}
	
	public void setPowerBlock(long chunk, int pos, PowerBlock block) {
		indexWireGrid(chunk)[pos] = block;
	}
	
	@Nullable
	public PowerBlock getPowerBlock(long chunk, int pos) {
		return indexWireGrid(chunk)[pos];
	}
	
	public PowerBlock getPowerBlock(BlockPosition blockPosition) {
		int chunkX = ChunkUtils.getChunkCoordinate(blockPosition.getX());
		int chunkZ = ChunkUtils.getChunkCoordinate(blockPosition.getZ());
		long chunk = ChunkUtils.getChunkIndex(chunkX, chunkZ);
		int positionIndex = ChunkUtils.getBlockIndex(blockPosition.getX(), blockPosition.getY(), blockPosition.getZ());
		
		return indexWireGrid(chunk)[positionIndex];
	}
	
	private PowerBlock[] indexWireGrid(long chunk) {
		PowerBlock[] chunkGrid = powerGrid.get(chunk);
		
		if (chunkGrid != null)
			return chunkGrid;
		
		SkylaExtension.LOGGER.info("Creating new power grid for chunk: " + chunk);
		
		chunkGrid = new PowerBlock[16 * 256 * 16];
		
		powerGrid.put(chunk, chunkGrid);
		
		return chunkGrid;
	}

	/**
	 * Registers a block
	 */
	public void registerBlock(long chunk, int pos, PowerBlock block) {
		indexWireGrid(chunk)[pos] = block;
	}
	
	/**
	 * Unregisters a block
	 */
	public void unregisterBlock(long chunk, int pos, PowerBlock block) {
		indexWireGrid(chunk)[pos] = null;
	}
}

File: src/main/java/main/block/power/PowerSystem.kt
package main.block.power

import it.unimi.dsi.fastutil.longs.Long2ObjectMap
import it.unimi.dsi.fastutil.longs.Long2ObjectOpenHashMap
import main.SkylaExtension
import net.minestom.server.instance.Instance
import net.minestom.server.utils.BlockPosition
import net.minestom.server.utils.chunk.ChunkUtils
import java.util.*

// Defines the behavior of the custom power system for one instance
class PowerSystem private constructor(val instance: Instance) {
    private val powerGrid: Long2ObjectMap<Array<PowerBlock?>> = Long2ObjectOpenHashMap()
    fun setPowerBlock(chunk: Long, pos: Int, block: PowerBlock?) {
        indexWireGrid(chunk)[pos] = block
    }

    fun getPowerBlock(chunk: Long, pos: Int): PowerBlock? {
        return indexWireGrid(chunk)[pos]
    }

    fun getPowerBlock(blockPosition: BlockPosition): PowerBlock? {
        val chunkX = ChunkUtils.getChunkCoordinate(blockPosition.x.toDouble())
        val chunkZ = ChunkUtils.getChunkCoordinate(blockPosition.z.toDouble())
        val chunk = ChunkUtils.getChunkIndex(chunkX, chunkZ)
        val positionIndex = ChunkUtils.getBlockIndex(blockPosition.x, blockPosition.y, blockPosition.z)
        return indexWireGrid(chunk)[positionIndex]
    }

    private fun indexWireGrid(chunk: Long): Array<PowerBlock?> {
        var chunkGrid = powerGrid[chunk]
        if (chunkGrid != null) return chunkGrid
        SkylaExtension.LOGGER!!.info("Creating new power grid for chunk: $chunk")
        chunkGrid = arrayOfNulls(16 * 256 * 16)
        powerGrid[chunk] = chunkGrid
        return chunkGrid
    }

    /**
     * Registers a block
     */
    fun registerBlock(chunk: Long, pos: Int, block: PowerBlock?) {
        indexWireGrid(chunk)[pos] = block
    }

    /**
     * Unregisters a block
     */
    fun unregisterBlock(chunk: Long, pos: Int, block: PowerBlock?) {
        indexWireGrid(chunk)[pos] = null
    }

    companion object {
        private val powerSystems: MutableMap<Instance, PowerSystem> = HashMap()
        fun of(instance: Instance): PowerSystem {
            if (powerSystems.containsKey(instance)) return powerSystems[instance]!!
            val powerSystem = PowerSystem(instance)
            powerSystems[instance] = powerSystem
            return powerSystem
        }
    }
}